Signing prop page

mavasani opened the pull request:
Add support to show the Signing property page and implement `InterceptedProjectPropertiesProvider` (allows to write a single property value provider to validate and/or transform a property value)

Fixes #265

**NOTE:** This PR depends on bunch of source and test changes from @natidea's PR https://github.com/dotnet/roslyn-project-system/pull/283/ - I have just cloned these files for now. Once that PR has been merged in, I'll remove the cloned copies in my branch.

mavasani commented on src/Microsoft.VisualStudio.AppDesigner/PropPages/PropertyControlData.vb:
```diff
@@ -1726,6 +1726,9 @@ Namespace Microsoft.VisualStudio.Editors.PropertyPages
                     Descriptor.SetValue(Component, Value)
 
+                    ' Workaround until CPS is fixed to fire the value changed event on Descriptor.SetValue
+                    _valueChangedWasFired = Value.Equals(Descriptor.GetValue(Component))
```
Is this workaround still needed?

davkean replied:
Please file a bug against us so that we can remove this when CPS fixes this.

mavasani replied:
Filed https://github.com/dotnet/roslyn-project-system/issues/289

srivatsn replied:
The big has already been fixed. You can remove the workaround

mavasani replied:
I'd like to keep this workaround until we all move to a newer build (probably next preview?). I'll revert it then.

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/IProjectPropertiesFactory.cs:
```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Mostly a clone from Nat's PR, with some modifications to allow handling multiple properties and helpers to setup get/set simulatneously.

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkValueProvider.cs:
```diff
@@ -0,0 +1,39 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties
+{
+    [ExportInterceptingPropertyValueProvider]
+    internal sealed class TargetFrameworkValueProvider : InterceptingPropertyValueProviderBase
```
This is what each intercepting provider value provider would require to implement.

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/GlobalSuppressions.cs:
```diff
@@ -5,12 +5,4 @@
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "RS0006:Do not mix attributes from different versions of MEF", Justification = "<Pending>", Scope = "type", Target = "~T:Microsoft.VisualStudio.ProjectSystem.SpecialFileProviders.ResourcesFileSpecialFileProvider")]
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "RS0006:Do not mix attributes from different versions of MEF", Justification = "<Pending>", Scope = "type", Target = "~T:Microsoft.VisualStudio.ProjectSystem.SpecialFileProviders.SettingsFileSpecialFileProvider")]
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "RS0006:Do not mix attributes from different versions of MEF", Justification = "<Pending>", Scope = "type", Target = "~T:Microsoft.VisualStudio.ProjectSystem.Properties.OutputTypeEnumProvider")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "RS0006:Do not mix attributes from different versions of MEF", Justification = "<Pending>")]
```
I have added a project level suppression for this rule to avoid everyone having to add a suppression for their newly added exports.

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectPropertiesProviders/DelegatedProjectPropertiesBase.cs:
```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Direct clone from Nat's PR.

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectPropertiesProviders/DelegatedProjectPropertiesProviderBase.cs:
```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Direct clone from Nat's PR.

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectPropertiesProviders/InterceptedProjectPropertiesProvider.cs:
```diff
@@ -0,0 +1,48 @@
+    /// </summary>
+    [Export("ProjectFileWithInterception", typeof(IProjectPropertiesProvider))]
+    [Export(typeof(IProjectPropertiesProvider))]
+    [ExportMetadata("Name", "ProjectFileWithInterception")]
```
Anyone with a better name suggestion?

mavasani commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/FileNameUtilties.cs:
```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
I had to copy over all the file and path utilities from Roslyn for TargetFrameworkValueProvider. You can ignore the review for rest of the 4 files.

mavasani commented:
Tagging @dotnet/project-system for review

davkean commented on src/Microsoft.VisualStudio.AppDesigner/PropPages/PropPageUserControlBase.vb:
```diff
@@ -3337,7 +3341,19 @@ NextControl:
         ''' <returns></returns>
         ''' <remarks></remarks>
         Protected Function GetProjectPath() As String
-            Return CStr(GetCommonPropertyValueNative("FullPath")) 'CONSIDER: This won't work for all project types, e.g. ASP.NET when project path is a URL
```
I don't understand why we needed to change this - we should be shimming these.

srivatsn replied:
GetCommonPropertyValue tries to get the property from the ProjectProperties interface that vslanproj implements. Without COM aggregation, is there anything else we can do here?

mavasani replied:
`GetCommonPropertyValueNative("FullPath")` doesn't seem to work.

jinujoseph replied:
I have a need to do the same for my "Reference Path " Properties Page and i was doing this .

''' <summary>
        ''' Gets the project's path
        ''' First look in PropertyDescriptor of a common property if not found look in ObjectsPropertyDescriptorsArray
        ''' </summary>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Protected Function GetProjectPath() As String
            Dim PropertyName = "FullPath"
            Dim propertyDescriptor = GetCommonPropertyDescriptor(PropertyName)

            If propertyDescriptor Is Nothing Then
                propertyDescriptor = GetPropertyDescriptor(PropertyName)
            End If

            Return CStr(GetCommonPropertyValueNative(propertyDescriptor))
        End Function

davkean replied:
I don't understand - what does GetCommonPropertyValueNative have to do with VSLangProj? It calls through the OAProject, here: http://index/#Microsoft.VisualStudio.AppDesigner/PropPages/PropPageUserControlBase.vb,489

davkean replied:
ie we should add a property that returns FullPath.

mavasani replied:
Property already exists in [GeneralBrowseObject.xaml](https://github.com/dotnet/roslyn-project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/GeneralBrowseObject.xaml#L58)?

davkean replied:
BrowseObject != Project Properties, basically if you add it to ConfigurationGeneral, it should show up.

davkean replied:
Hmm, actually maybe I'm wrong - it is calling through VSLangProj, sigh. I can't tell what object that is, can you set a break point and tell me what DTEProject.Object returns?

srivatsn replied:
No DTE.Object returns OAVsProject which doesn't have return the properties from ConfigurationGeneral. http://index/#Microsoft.VisualStudio.ProjectSystem.VS.Implementation/Package/Automation/VSProject/OAVSProject.cs,0a10689b8d4b8823

mavasani replied:
I have filed a work item to clean this up once ComAggregation is supported by CPS - https://github.com/dotnet/roslyn-project-system/issues/291

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ExportInterceptingPropertyValueProviderAttribute.cs:
```diff
@@ -0,0 +1,23 @@
+    /// </summary>
+    [MetadataAttribute]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = false, Inherited = false)]
+    public sealed class ExportInterceptingPropertyValueProviderAttribute : ExportAttribute
```
What value does this add over [Export(typeof(IInterceptingPropertyValueProvider))]?

mavasani replied:
The attribute is to keep this consistent with [ExportDynamicEnumValuesProviderAttribute](http://index/Microsoft.VisualStudio.ProjectSystem/R/9268c241811fc976.html), which has a similar functionality. We should potentially consider moving this intercepting value provider down to the CPS layer, as this seems like generic functionality.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/FilePathUtilties.cs:
```diff
@@ -0,0 +1,111 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem.Utilities
+{
+    internal static class FilePathUtilities
```
This is used just to append the backslash? Please just use `PathHelper.EnsureTrailingSlash`.

mavasani replied:
I used this to get relative path in TargetFrameworkValueProvider, I wasn't aware of the existing utility method that you pointed out in another comment. I'll remove these files.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkValueProvider.cs:
```diff
@@ -0,0 +1,39 @@
+
+        public override async Task<string> OnGetEvaluatedPropertyValueAsync(string evaluatedPropertyValue, IProjectProperties defaultProperties)
+        {
+            Requires.NotNull(_projectVsServices.VsHierarchy, "vsHierarchy");
```
This is strange, why does it do this? How would the caller even understand if this threw? This is a bug in the implementor IUnconfiguredProjectVsServices, not the caller.

mavasani replied:
I wasn't sure about the contract here, I'll remove the assert.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Properties/InterceptedProjectProperties/InterceptedProjectPropertiesProviderTests.cs:
```diff
@@ -0,0 +1,75 @@
+    [ProjectSystemTrait]
+    public class InterceptedProjectPropertiesProviderTests
+    {
+        private class MockInterceptedProjectPropertyProvider : InterceptingPropertyValueProviderBase
```
Just use Moq for this - it does all of these.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkValueProviderTests.cs:
```diff
@@ -0,0 +1,77 @@
+        {
+            var configuredTargetFramework = (uint)0x50000;
+            var properties = GetMockInterceptedProperties(configureHierarchy: true, configuredTargetFramework: configuredTargetFramework);
+            var propertyValueStr = properties.GetEvaluatedPropertyValueAsync(TargetFrameworkPropertyName).Result;
```
You use task on every test except here - can you be consistent?

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkValueProvider.cs:
```diff
@@ -0,0 +1,39 @@
+            object targetFrameworkVersion;
+            if (ErrorHandler.Succeeded(_projectVsServices.VsHierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)Shell.VsHierarchyPropID.TargetFrameworkVersion, out targetFrameworkVersion)))
+            {
+                return ((uint)targetFrameworkVersion).ToString();
```
We shouldn't be going through the VsHierarchy for a couple of reasons:

1) This type will be called on background threads, IVsHierarchy is UI-bound.
2) You should be using the accessing the property itself - see ManagedDebuggerImageTypeService for an example for reading a property. IVsHierarchy is for legacy consumers.

mavasani replied:
> This type will be called on background threads, IVsHierarchy is UI-bound.

I thought IVsUIHierarchy was UI bound, not IVsHierarchy. I initially tried implementing this through the UI hierarchy and it indeed deadlocked because this provider is invoked from the backgrond thread while the main thread is waiting on it the property value. Switching to `IUnconfiguredProjectVsServices.VsHierarchy` actually fixed the deadlock as CPS implements this without switching to the UI thread. It hits [this](http://index/#Microsoft.VisualStudio.ProjectSystem.VS.Implementation/Package/Nodes/ProjectNode.Hierarchy.cs,761) code path to compute it.

> You should be using the accessing the property itself - see ManagedDebuggerImageTypeService for an example for reading a property.

There is no property named "TargetFramework" on [ConfigurationGeneral](http://index/Microsoft.VisualStudio.ProjectSystem.VS.Implementation/R/cc66d9dab353fda3.html). CPS implements this by fetching the "TargetFrameworkMoniker" property and performing [this](http://index/#Microsoft.VisualStudio.ProjectSystem.VS.Implementation/Package/Nodes/ProjectNode.Hierarchy.cs,765) mapping.

srivatsn replied:
You can add a definition for TargetFramework as a StringProperty to ConfigurationGeneral.xaml, since that's just a MSBuild property and that will then show up on that type.

mavasani replied:
Adding it to ConfigurationGeneral will ensure it shows up on that type, but how will it be implemented? There is no MSBuild property named "TargetFramework" in the project file.

TargetFramework was already non-visible a string property defined in [GeneralBrowseObject.xaml](https://github.com/dotnet/roslyn-project-system/pull/287/files#diff-a7ea57aece378dbf2944a9c60361a729L29). I had to change it to IntProperty as SigningPropPage (and potentially other consumers) expect this to be integral.

srivatsn replied:
There's a TargetFrameworkVersion property in the targets. TargetFrameworkMoniker is defined in terms of that property - [see](http://index/#MSBuildFiles/C/ProgramFiles(x86)/MSBuild/14.0/bin_/amd64/Microsoft.Common.CurrentVersion.targets,75)

davkean replied:
All IVsXXX interfaces are UI-bound.

mavasani replied:
Yes, but even that is not an integral value. Its of the format `<TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>`. "TargetFramework" seems to be a derived property created by taking the moniker and then performing [bit manipulation](http://index/#Microsoft.VisualStudio.ProjectSystem.VS.Implementation/Package/Nodes/ProjectNode.Hierarchy.cs,765) with the framework version - CPS already has an API on IUnconfiguredVsProjectServices.VsHierarchy to fetch this evaluated value in a non-UI thread way, which I have used here.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/IInterceptingPropertyValueProvider.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        /// Gets the name of the property.
+        /// </summary>
+        string GetPropertyName();
```
This should be metadata value not a method, so that you don't instantiate these things unless actually needed.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/InterceptedProjectProperties.cs:
```diff
@@ -0,0 +1,72 @@
+            var builder = ImmutableDictionary.CreateBuilder<string, IInterceptingPropertyValueProvider>(StringComparer.OrdinalIgnoreCase);
+            foreach (var valueProvider in valueProviders)
+            {
+                var properyName = valueProvider.Value.GetPropertyName();
```
If this is a metadata name, then you don't need to instantiate these unless absolutely needed.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/SigningPropertyPage/AssemblyOriginatorKeyFileValueProvider.cs:
```diff
@@ -0,0 +1,35 @@
+            try
+            {
+                var projectFileFullPath = defaultProperties.FileFullPath;
+                if (!string.IsNullOrEmpty(projectFileFullPath) && PathUtilities.IsAbsolute(unevaluatedPropertyValue))
```
Instead of copying Roslyn's code here, can we just use existing proejct system code, in particular, PathHelper.TryMakeRelativeToProjectDirectory and PathHelper.XXXRooted API?

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/SigningPropertyPage/AssemblyOriginatorKeyFileValueProvider.cs:
```diff
@@ -0,0 +1,35 @@
+                    unevaluatedPropertyValue = FilePathUtilities.GetRelativePath(projectDirectory, unevaluatedPropertyValue);
+                }
+            }
+            catch(Exception)
```
This can't be the correct behavior, someone sets a bad path and we ignore it?

mavasani replied:
I am not sure about the contract here - is each IProjectPropertiesProvider responsible for handling all exceptions or something down the stack will handle them? We surely don't want to crash the process due to a bad path.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/GeneralBrowseObject.xaml:
```diff
@@ -26,11 +26,11 @@
+  <IntProperty Name="TargetFramework" Visible="False">
+    <IntProperty.DataSource>
+      <DataSource Persistence="ProjectFileWithInterception" PersistedName="TargetFramework" />
+    </IntProperty.DataSource>
```
I think we'll want to make this read-only as well. This should never be persisted.

mavasani commented:
I have pushed a commit addressing the PR feedback.

Additionally, can someone take a look at VSO PR for a trivial fix in SigningPropPage [here](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/27778?view=compare&path=%2Fsrc%2Fwizard%2Fvbdesigner%2FMicrosoft.VisualStudio.Editors.ClickOnce%2FPropPages%2FSigningPropPage.vb)

mavasani commented:
@dotnet-bot retest this please

mavasani commented:
@dotnet/project-system I have merged latest sources and deleted the cloned code. Can you please take another look so I can unblock @srivatsn ?

mavasani commented:
Ping?

srivatsn commented:
👍
